src/UCX.jl

```diff
@@ -254,52 +276,73 @@ and call callbacks.
 Returns `true` if progress was made, `false` if no work was waiting.
 """
 function progress(worker::UCXWorker)
-    API.ucp_worker_progress(worker) != 0
+    tid = Base.Threads.threadid()
+    if worker.in_amhandler[tid]
+        @debug """
+        UCXWorker is processing a Active Message on this thread
+        Calling `progress` is not permitted and leads to recursion.
+        """ tid exception=(UCXException(API.UCS_ERR_NO_PROGRESS), catch_backtrace())
+        return false
+    else
+        return API.ucp_worker_progress(worker) != 0
+    end
 end
 
 function fence(worker::UCXWorker)
     @check API.ucp_worker_fence(worker)
 end
 
+include("idle.jl")
+
 import FileWatching: poll_fd
 function Base.wait(worker::UCXWorker)
-    # Approach 0: Timer
-    # Turns out to be very slow
-    # Approach 2: Busy loop
-    # Unfair scheduling :/
-    # while progress(worker) && isopen(worker)
-    #     yield()
-    # end
-    # Approach 3
-    # About 2x slower on small messages than busy
-    while isopen(worker)
-        if progress(worker)
-            # progress was made
-            yield()
-            continue
+    if WORKER_IDLE[]
+        idler = UvWorkerIdle(worker)
+        wait(idler)
+        close(idler)
+    elseif WORKER_BUSY[]
+        while isopen(worker)
+            # Temporary solution before we have gc transition support in codegen.
+            ccall(:jl_gc_safepoint, Cvoid, ())
```

Comment:
@maleadt this is the loop that livelocked without a safepoint, .... maybe progress always returned false.